share/man/man8/man8.vax/format.8
.\" Copyright (c) 1980, 1983 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	from: @(#)format.8	6.6 (Berkeley) 4/20/91
.\"	$Id: format.8,v 1.1 1995/10/18 08:44:41 deraadt Exp $
.\"
.TH FORMAT 8 "April 20, 1991"
.UC 4
.SH NAME
format \- how to format disk packs
.SH DESCRIPTION
.PP
There are two ways to format disk packs.  The simplest is to
use the
.I format
program.  The alternative is to use the DEC standard formatting
software which operates under the DEC diagnostic supervisor. 
This manual page describes the operation of
.IR format ,
then concludes with some remarks about using the DEC formatter.
.PP
.I Format
is a standalone program used to
format and check disks prior to constructing
file systems.
In addition to the formatting
operation,
.I format
records any bad sectors encountered
according to DEC standard 144. 
Formatting is performed one track at a
time by writing the appropriate
headers and a test pattern and then checking
the sector by reading and verifying the pattern, using
the controller's ECC for error detection.
A sector is marked bad if an unrecoverable media
error is detected, or if a correctable
ECC error too many bits in length
is detected (such errors are indicated as
``ECC'' in the summary printed upon
completing the format operation).
After the entire disk
has been formatted and checked,
the total number of errors are reported,
any bad sectors and skip sectors are marked,
and a bad sector forwarding table
is written to the disk in the first five
even numbered sectors of the last track.
It is also possible to reformat sections of the disk
in units of tracks.
.I Format 
may be used on any UNIBUS or MASSBUS drive
supported by the \fIup\fP and \fIhp\fP device
drivers which uses 4-byte headers (everything except RP's).
.PP
The test pattern used during the media check
may be selected from one of: 0xf00f (RH750 worst case),
0xec6d (media worst case), and 0xa5a5 (alternating
1's and 0's).
Normally the media worst case pattern is used.
.PP
.I Format
also has an option to perform an extended \*(lqsevere burn-in,\*(rq
which makes a number of passes using different patterns.
The number of passes can be selected at run time,
up to a maximum of 48,
with provision for additional passes or termination
after the preselected number of passes.
This test runs for many hours, depending on the disk and processor.
.PP
Each time
.I format
is run to format an entire disk,
a completely new bad sector table is generated
based on errors encountered while formatting.
The device driver, however, will always attempt to
read any existing bad sector table when the device is
first opened.
Thus, if a disk pack has never previously been formatted,
or has been formatted with different sectoring,
five error messages will be printed when the driver attempts
to read the bad sector table; these diagnostics should be ignored.
.PP
Formatting a 400 megabyte disk on a MASSBUS disk controller
usually takes about 20 minutes.
Formatting on a UNIBUS disk controller takes significantly
longer.
For every hundredth cylinder formatted
.I format
prints a message indicating the current cylinder being formatted.
(This message is just to reassure people that nothing is
is amiss.)
.PP
.I Format
uses the standard notation of the standalone I/O library
in identifying a drive to be formatted.  A drive is
specified as
.IR zz ( x , y ),
where 
.I zz
refers to the controller type (either
.I hp
or
.IR up ),
.I x
is the unit number of the drive;
8 times the UNIBUS or MASSBUS
adaptor number plus the MASSBUS drive number or UNIBUS
drive unit number; and
.I y
is the file system partition
on drive
.I x
(this should always be 0). 
For example, ``hp(1,0)'' indicates that drive
1 on MASSBUS adaptor 0 should be formatted; while
``up(10,0)'' indicates that UNIBUS drive 2 on UNIBUS adaptor 1
should be formatted.
.PP
Before each formatting attempt, 
.I format
prompts the user in case debugging should be enabled
in the appropriate device driver.  A carriage return
disables debugging information.
.PP
.I Format
should be used prior to building file systems (with 
.IR newfs (8))
to insure that all sectors with uncorrectable media errors
are remapped.  If a drive develops uncorrectable
defects after formatting, either
.IR bad144 (8)
or
.IR badsect (8)
should be able to avoid the bad sectors.
.SH EXAMPLE
A sample run of 
.I format
is shown below.  
In this example (using a VAX-11/780), 
.I format
is loaded from the console floppy;
on an 11/750
.I format
will be loaded from the root file system with
.IR boot (8)
following a \*(lqB/3\*(rq command.
Boldface means user input.
As usual, ``#'' and ``@'' may be used to edit input.
.nf
.in +0.5i
.ta  1i
.sp 1
>>>\fBL FORMAT\fP
	LOAD DONE, 00004400 BYTES LOADED
>>>\fBS 2\fP
Disk format/check utility
.sp
Enable debugging (0=none, 1=bse, 2=ecc, 3=bse+ecc)? \fB0\fP
Device to format? \fBhp(8,0)\fP
(\fIerror messages may occur as old bad sector table is read\fP)
Formatting drive hp0 on adaptor 1: verify (yes/no)? \fByes\fP
Device data: #cylinders=842, #tracks=20, #sectors=48
Starting cylinder (0):
Starting track (0):
Ending cylinder (841):
Ending track (19):
Available test patterns are:
.in +1.0i
1 - (f00f) RH750 worst case
2 - (ec6d) media worst case
3 - (a5a5) alternating 1's and 0's
4 - (ffff) Severe burnin (up to 48 passes)
.in -1.0i
Pattern (one of the above, other to restart)? \fB2\fP
Maximum number of bit errors to allow for soft ECC (3):
Start formatting...make sure the drive is online
 ...
(\fIsoft ecc's and other errors are reported as they occur\fP)
 ...
(\fIif 4 write check errors were found, the program terminates like this...\fP)
 ...
Errors:
Bad sector: 0
Write check: 4
Hard ECC: 0
Other hard: 0
Marked bad: 0
Skipped: 0
Total of 4 hard errors revectored.
Writing bad sector table at block 808272	
(\fI808272 is the block # of the first block in the bad sector table\fP)
Done
(\fI...program restarts to allow formatting other disks\fP)
(\fI...to abort halt machine with ^P\fP)
.fi
.sp 1
.SH DIAGNOSTICS
The diagnostics are intended to be self explanatory.
.SH "USING DEC SOFTWARE TO FORMAT"
.PP
.B "Warning:  These instructions are for people with 11/780 CPU's."
The steps needed for 11/750 or 11/730 cpu's are similar, but not
covered in detail here.
.I
.PP
The formatting procedures are different for each type of disk.  Listed
here are the formatting procedures for RK07's, RP0X, and RM0X 
disks.
.PP
You should shut down UNIX and halt the machine to do any disk formatting.
Make certain you put in the pack you want formatted.  It is also a good idea
to spin down or write protect the disks you don't want to format, just in case.
.PP
.B "Formatting an RK07."
Load the console floppy labeled, "RX11 VAX DSK LD DEV #1" in
the console disk drive, and type the following commands:
.RS
.nf
>>>BOOT
DIAGNOSTIC SUPERVISOR.  ZZ-ESSAA-X5.0-119  23-JAN-1980 12:44:40.03
DS>ATTACH DW780 SBI DW0 3 5
DS>ATTACH RK611 DMA
DS>ATTACH RK07 DW0 DMA0
DS>SELECT DMA0
DS>LOAD EVRAC
DS>START/SEC:PACKINIT
.fi
.RE
.PP
.B "Formatting an RP0X."
Follow the above procedures except that the
ATTACH and SELECT lines should read:
.RS
.nf
DS>ATTACH RH780 SBI RH0 8 5
DS>ATTACH RP0X RH0 DBA0		(RP0X is, e.g. RP06)
DS>SELECT DBA0
.fi
.RE
.PP
This is for drive 0 on mba0; use 9 instead of 8 for mba1, etc.
.PP
.B "Formatting an RM0X."
Follow the above procedures except that the
ATTACH and SELECT lines should read:
.RS
.nf
DS>ATTACH RH780 SBI RH0 8 5
DS>ATTACH RM0X RH0 DRA0
DS>SELECT DRA0
.fi
.RE
.PP
Don't forget to put your UNIX console floppy back
in the floppy disk drive.
.SH SEE ALSO
bad144(8),
badsect(8),
newfs(8)
.SH BUGS
An equivalent facility should be available which operates under
a running UNIX system. 
.PP
It should be possible to reformat or verify part or all of a disk,
then update the existing bad sector table.


gnu/usr.bin/gdb/bfd/format.c
/* Generic BFD support for file formats.
   Copyright (C) 1990-1991 Free Software Foundation, Inc.
   Written by Cygnus Support.

This file is part of BFD, the Binary File Descriptor library.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

	$Id: format.c,v 1.1 1995/10/18 08:39:53 deraadt Exp $
*/

/*
SECTION
	File Formats

	A format is a BFD concept of high level file contents. The
	formats supported by BFD are: 

	o bfd_object

	The BFD may contain data, symbols, relocations and debug info.

	o bfd_archive

	The BFD contains other BFDs and an optional index.

	o bfd_core

	The BFD contains the result of an executable core dump.


*/

#include "bfd.h"
#include "sysdep.h"
#include "libbfd.h"

extern bfd_target *target_vector[];
extern bfd_target *default_vector[];


/*
FUNCTION
	bfd_check_format

SYNOPSIS
	boolean bfd_check_format(bfd *abfd, bfd_format format);

DESCRIPTION
	This routine is supplied a BFD and a format. It attempts to
	verify if the file attached to the BFD is indeed compatible
	with the format specified (ie, one of <<bfd_object>>,
	<<bfd_archive>> or <<bfd_core>>).

	If the BFD has been set to a specific @var{target} before the
	call, only the named target and format combination will be
	checked. If the target has not been set, or has been set to
	<<default>> then all the known target backends will be
	interrogated to determine a match.  If the default target
	matches, it is used.  If not, exactly one target must recognize
	the file, or an error results.

	The function returns <<true>> on success, otherwise <<false>>
	with one of the following error codes:  

	o invalid_operation -
	if <<format>> is not one of <<bfd_object>>, <<bfd_archive>> or
	<<bfd_core>>.

	o system_call_error -
	if an error occured during a read - even some file mismatches
	can cause system_call_errors

	o file_not_recognised -
	none of the backends recognised the file format

	o file_ambiguously_recognized -
	more than one backend recognised the file format.

*/

boolean
DEFUN(bfd_check_format,(abfd, format),
      bfd *abfd AND
      bfd_format format)
{
  bfd_target **target, *save_targ, *right_targ;
  int match_count;

  if (!bfd_read_p (abfd) ||
      ((int)(abfd->format) < (int)bfd_unknown) ||
      ((int)(abfd->format) >= (int)bfd_type_end)) {
    bfd_error = invalid_operation;
    return false;
  }

  if (abfd->format != bfd_unknown)
    return (abfd->format == format)? true: false;


  /* Since the target type was defaulted, check them 
     all in the hope that one will be uniquely recognized.  */

  save_targ = abfd->xvec;
  match_count = 0;
  right_targ = 0;


  /* presume the answer is yes */
  abfd->format = format;

  /* If the target type was explicitly specified, just check that target.  */

  if (!abfd->target_defaulted) {
    bfd_seek (abfd, (file_ptr)0, SEEK_SET);	/* rewind! */

    right_targ = BFD_SEND_FMT (abfd, _bfd_check_format, (abfd));
    if (right_targ) {
      abfd->xvec = right_targ;		/* Set the target as returned */
      return true;			/* File position has moved, BTW */
    }
  }

  for (target = target_vector; *target != NULL; target++) {
    bfd_target *temp;

    abfd->xvec = *target;	/* Change BFD's target temporarily */
    bfd_seek (abfd, (file_ptr)0, SEEK_SET);
    /* If _bfd_check_format neglects to set bfd_error, assume wrong_format.
       We didn't used to even pay any attention to bfd_error, so I suspect
       that some _bfd_check_format might have this problem.  */
    bfd_error = wrong_format;
    temp = BFD_SEND_FMT (abfd, _bfd_check_format, (abfd));
    if (temp) {				/* This format checks out as ok! */
      right_targ = temp;
      match_count++;
      /* If this is the default target, accept it, even if other targets
	 might match.  People who want those other targets have to set 
	 the GNUTARGET variable.  */
      if (temp == default_vector[0])
	{
	  match_count = 1;
	  break;
	}
#ifdef GNU960
      /* Big- and little-endian b.out archives look the same, but it doesn't
       * matter: there is no difference in their headers, and member file byte
       * orders will (I hope) be handled appropriately by bfd.  Ditto for big
       * and little coff archives.  And the 4 coff/b.out object formats are
       * unambiguous.  So accept the first match we find.
       */
      break;
#endif
    } else if (bfd_error != wrong_format) {
      abfd->xvec = save_targ;
      abfd->format = bfd_unknown;
      return false;
    }
  }

  if (match_count == 1) {
    abfd->xvec = right_targ;		/* Change BFD's target permanently */
    return true;			/* File position has moved, BTW */
  }

  abfd->xvec = save_targ;		/* Restore original target type */
  abfd->format = bfd_unknown;		/* Restore original format */
  bfd_error = ((match_count == 0) ? file_not_recognized :
	       file_ambiguously_recognized);
  return false;
}


/*
FUNCTION
	bfd_set_format

SYNOPSIS
	boolean bfd_set_format(bfd *, bfd_format);

DESCRIPTION
	This function sets the file format of the supplied BFD to the
	format requested. If the target set in the BFD does not
	support the format requested, the format is illegal or the BFD
	is not open for writing than an error occurs.

*/

boolean
DEFUN(bfd_set_format,(abfd, format),
      bfd *abfd AND
      bfd_format format)
{

  if (bfd_read_p (abfd) ||
      ((int)abfd->format < (int)bfd_unknown) ||
      ((int)abfd->format >= (int)bfd_type_end)) {
    bfd_error = invalid_operation;
    return false;
  }

  if (abfd->format != bfd_unknown)
    return (abfd->format == format) ? true:false;

  /* presume the answer is yes */
  abfd->format = format;

  if (!BFD_SEND_FMT (abfd, _bfd_set_format, (abfd))) {
    abfd->format = bfd_unknown;
    return false;
  }

  return true;
}


/*
FUNCTION
	bfd_format_string

SYNOPSIS
	CONST char *bfd_format_string(bfd_format);

DESCRIPTION
	This function takes one argument, and enumerated type
	(bfd_format) and returns a pointer to a const string
	<<invalid>>, <<object>>, <<archive>>, <<core>> or <<unknown>>
	depending upon the value of the enumeration.
*/

CONST char *
DEFUN(bfd_format_string,(format),
     bfd_format format)
{
  if (((int)format <(int) bfd_unknown) 
      || ((int)format >=(int) bfd_type_end)) 
    return "invalid";
  
  switch (format) {
  case bfd_object:
    return "object";		/* linker/assember/compiler output */
  case bfd_archive: 
    return "archive";		/* object archive file */
  case bfd_core: 
    return "core";		/* core dump */
  default: 
    return "unknown";
  }
}
